### PR TITLE
Remove dead code and cleanup scoping in dmu_send.c

### PR DESCRIPTION
**Pull request by mattmacy**
This addresses a number of problems with dmu_send.c:

- `bp_span` is unused which makes clang complain
- `dump_write` conflicts with FreeBSD's existing core dump code
- `range_alloc` is private to the file and not declared in any headers causing clang to complain
- `dump_spill` and `arc_buf_destroy` should not be called if `dso_dryrun` is set

Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/dmu_send.c [hunk 1]
```diff
@@ -959,9 +948,11 @@ do_dump(dmu_send_cookie_t *dscp, struct send_range *range)
...
 			    zioflags, &aflags, &zb) != 0)
 				return (SET_ERROR(EIO));
-
-			err = dump_spill(dscp, bp, zb.zb_object, abuf->b_data);
-			arc_buf_destroy(abuf, &abuf);
+			if (!dscp->dsc_dso->dso_dryrun) {
+				err = dump_spill(dscp, bp, zb.zb_object,
+				    abuf->b_data);
+				arc_buf_destroy(abuf, &abuf);
+			}
```
Comment (ahrens) [inline module/zfs/dmu_send.c]: It looks like this is a bug fix, which we might want to backport to 0.8.  Could you elaborate on how we hit this, and what the impact is?

Comment (behlendorf) [reply module/zfs/dmu_send.c]: I'd suggest opening a PR with just this fix to make it easy to backport.

Comment (mattmacy) [reply module/zfs/dmu_send.c]: It was an issue I hit when doing the initial porting work. I'll just revert it and make a mental note.

Comment (mattmacy) [reply module/zfs/dmu_send.c]: If you look at the code it seems self evident that we're dumping a buffer and destroying an arc_buf when no read has happened: 
```
			arc_buf_t *abuf;
			zbookmark_phys_t zb;
			ASSERT3U(range->start_blkid, ==, DMU_SPILL_BLKID);
			zb.zb_objset = dmu_objset_id(dscp->dsc_os);
			zb.zb_object = range->object;
			zb.zb_level = 0;
			zb.zb_blkid = range->start_blkid;

			if (!dscp->dsc_dso->dso_dryrun && arc_read(NULL, spa,
			    bp, arc_getbuf_func, &abuf, ZIO_PRIORITY_ASYNC_READ,
			    zioflags, &aflags, &zb) != 0)
				return (SET_ERROR(EIO));
			if (!dscp->dsc_dso->dso_dryrun) {
				err = dump_spill(dscp, bp, zb.zb_object,
				    abuf->b_data);
				arc_buf_destroy(abuf, &abuf);
			}
```

File: module/zfs/dmu_send.c [hunk 2]
```diff
@@ -959,7 +948,6 @@ do_dump(dmu_send_cookie_t *dscp, struct send_range *range)
 			    bp, arc_getbuf_func, &abuf, ZIO_PRIORITY_ASYNC_READ,
 			    zioflags, &aflags, &zb) != 0)
 				return (SET_ERROR(EIO));
-
```
Comment (behlendorf) [inline module/zfs/dmu_send.c]: nit: let's keep the white space here.

File: module/zfs/dmu_send.c [hunk 3]
```diff
@@ -1076,7 +1064,8 @@ do_dump(dmu_send_cookie_t *dscp, struct send_range *range)
 			} else {
 				psize = BP_GET_PSIZE(bp);
 			}
-			err = dump_write(dscp, srdp->obj_type, range->object,
+			err = dmu_dump_write(dscp, srdp->obj_type,
+			    range->object,
```
Comment (behlendorf) [inline module/zfs/dmu_send.c]: `offset, srdp->datablksz, psize, bp,` can all be added to this line and still be under the limit.

**Pull request merged**